src/ipa_pc/mod.rs

```diff
@@ -560,24 +767,102 @@ impl<G: AffineCurve, D: Digest> PolynomialCommitment<G::ScalarField> for InnerPr
                     );
-                    combined_rand += &(cur_challenge * &shifted_rand.unwrap());
+                    let comm_single = commitment.comm[i];
+                    polynomial_lc += (point.pow(&[is as u64]), &poly_single);
+                    comm_lc += &comm_single.mul(point.pow(&[is as u64]));
+                    if has_hiding {
+                        let rand_single = randomness.rand[i];
+                        rand_lc += &(point.pow(&[is as u64]) * rand_single);
+                    }
+                }
+
+                combined_polynomial += (cur_challenge, &polynomial_lc);
+                combined_commitment_proj += &comm_lc.mul(&cur_challenge);
+                if has_hiding {
+                    combined_rand += &(cur_challenge * &rand_lc);
+                }
+
+            } else {
+
+                combined_polynomial += (cur_challenge, polynomial);
+                combined_commitment_proj += &commitment.comm[0].mul(cur_challenge);
+
+                if has_hiding {
+                    combined_rand += &(cur_challenge * &randomness.rand[0]);
                 }
             }
 
             cur_challenge = opening_challenges(opening_challenge_counter);
             opening_challenge_counter += 1;
+
+            if let Some(degree_bound_len) = degree_bound_len {
+
+                if segments_count > 1 {
+
+                    let shifted_degree_bound = degree_bound_len % key_len - 1;
+                    let last_segment_polynomial = Polynomial::from_coefficients_slice(
+                        &polynomial.coeffs[(segments_count - 1) * key_len..p_len]
+                    );
+                    let shifted_polynomial = Self::shift_polynomial(
+                        ck,
+                        &last_segment_polynomial,
+                        shifted_degree_bound
+                    );
+                    let shift = -point.pow(&[(key_len - shifted_degree_bound - 1) as u64]);
+
+                    combined_polynomial += (cur_challenge, &shifted_polynomial);
+                    combined_commitment_proj += &commitment.shifted_comm.unwrap().mul(cur_challenge);
+
+                    combined_polynomial += (cur_challenge * &shift, &last_segment_polynomial);
+                    combined_commitment_proj += &commitment.comm[segments_count - 1].mul(cur_challenge * &shift);
+
+                    if hiding_bound.is_some() {
+                        let shifted_rand = randomness.shifted_rand;
+                        assert!(
+                            shifted_rand.is_some(),
+                            "shifted_rand.is_none() for {}",
+                            label
+                        );
+                        combined_rand += &(cur_challenge * &shifted_rand.unwrap());
+                        combined_rand += &(cur_challenge * &shift * &randomness.rand[segments_count - 1]);
+                    }
+
+                } else {
+
```

Comment:
Although it is more efficient to treat single segment sized polynomials differently, it is hard to understand that from the code :) I'd recommend to add a comment here that explains the different treatment.

Reply:
All parts of code where we have different branches for segmented and none segmented cases are removed